lib/client.js

```diff
@@ -2096,6 +2162,125 @@ MatrixClient.prototype.searchMessageText = function(opts, callback) {
+                filter: opts.filter,
+                order_by: "recent",
+                event_context: {
+                    before_limit: 1,
+                    after_limit: 1,
+                    include_profile: true,
+                }
+            }
+        }
+    };
+
+    var searchResults = {
+        _query: body,
+        results: [],
+        highlights: [],
+    };
+
+    return this.search({body: body}).then(
+        this._processRoomEventsSearch.bind(this, searchResults)
+    );
+};
+
+/**
+ * Take a result from an earlier searchRoomEvents call, and backfill results.
+ *
+ * @param  {module:models/search-results.SearchResults} searchResults   the
+ *    results object to be updated
+ * @return {module:client.Promise} Resolves: updated result object
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.backPaginateRoomEventsSearch = function(searchResults) {
+    // TODO: we should implement a backoff (as per scrollback()) to deal more
+    // nicely with HTTP errors.
+
+    if (!searchResults.next_batch) {
+        throw new Error("Cannot backpaginate event search any further");
+    }
+
+    if (searchResults.pendingRequest) {
+        // already a request in progress - return the existing promise
+        return searchResults.pendingRequest;
+    }
+
+    var searchOpts = {
+        body: searchResults._query,
+        next_batch: searchResults.next_batch,
+    };
+
+    var promise = this.search(searchOpts).then(
+        this._processRoomEventsSearch.bind(this, searchResults)
+    ).finally(function() {
+        searchResults.pendingRequest = null;
+    });
+    if (promise.isPending()) {
+        searchResults.pendingRequest = promise;
+    }
+
+    return promise;
+};
+
+/**
+ * helper for searchRoomEvents and backPaginateRoomEventsSearch. Processes the
+ * response from the API call and updates the searchResults
+ *
+ * @param {module:models/search-results.SearchResults} searchResults
+ * @param {Object} response
+ * @return {module:models/search-results.SearchResults} searchResults
+ * @private
+ */
+MatrixClient.prototype._processRoomEventsSearch = function(searchResults, response) {
+    var room_events = response.search_categories.room_events;
+
+    searchResults.count = room_events.count;
+    searchResults.next_batch = room_events.next_batch;
+
+    // combine the highlight list with our existing list; build an object
+    // to avoid O(N^2) fail
+    var highlights = {};
+    room_events.highlights.forEach(function(hl) { highlights[hl] = 1; });
+    searchResults.highlights.forEach(function(hl) { highlights[hl] = 1; });
+
+    // turn it back into a list.
+    searchResults.highlights = Object.keys(highlights);
+
+    // append the new results to our existing results
+    for (var i = 0; i < room_events.results.length; i++) {
+        var sr = SearchResult.fromJson(room_events.results[i], this.getEventMapper());
+        searchResults.results.push(sr);
```

Comment:
ugh, this is actually a failure of the jsdoc on _processRoomEventsSearch. It doesn't take a SearchResults (there is no such thing) but rather an anonymous object (with the fields as doucmented on searchRoomEvents).